---
output:
html_document:
toc: true                               # table of contents
toc_float: true                         # float the table of contents to the left of the main document content
toc_depth: 3                            # header levels 1,2,3
theme: default
number_sections: false                  # add section numbering to headers
df_print: paged                         # tables are printed as an html table with support for pagination over rows and columns
highlight: pygments
pdf_document: true
html_notebook:
  toc: yes
---


---
title: "Proteome-transcriptome-overlaps"
subtitle: ""
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: html_document
---

```{r loading-libraries, include=FALSE}
library(dplyr)
library(stringr)
library(ggplot2)
library(data.table)
#library(ComplexUpset)
library(UpSetR)
library(ggpubr)

theme_set(theme_bw() + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()))

options(error = function() traceback())

outdir = "results"
dir.create(outdir)
```

# Loading tables

Loading tables with patient peptide protein matches, and trasncriptomics table with BCR sequences from all patients.

The peptide-protein match table, with annotated CDR - FWR region was generated by running the `peptide_matches_CDR3.R` script on the raw proteomics data as well as the IgBlast transcriptomics data, providing the amino acid sequences of the BCRs and their regions.

```{r loading-tables, include=FALSE}
tx <- read.table("../BCR_analysis/repertoire_per_patient/repertoire_comparison/all_data.tsv", sep = "\t", header = T)
px <- read.table(paste0(outdir,"all_patients_peptide_protein_match_region_merged_tables.tsv"), sep = "\t", header = T)

px$time_point <- factor(px$time_point, levels=c("baseline", "6months", "12months"))
```

# Match distribution before filtering

Number of proteome peptide matchings per time point to the transcriptome before multimatch filtering:

```{r seq_numbers_plot_unique_before, fig.width=10, fig.asp=0.3, echo=FALSE}
px_peptides <- px %>% dplyr::select(peptide_id, time_point, subject_id) %>%
                      dplyr::distinct()

ggplot(px_peptides, aes(x=subject_id, fill=time_point)) +
  geom_bar(position = position_dodge()) +
  ggtitle("Before filtering")
ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_before_multimatch_filtering.pdf"), width = 15, height = 6, units = "cm")
ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_before_multimatch_filtering.png"), width = 15, height = 6, units = "cm")
```

Same plot as boxplot with counts per patient:

```{r seq_numbers_plot_unique, fig.width=3, fig.asp=1, echo=FALSE}
px_tp_sbj <- px %>% dplyr::select(peptide_id, time_point, subject_id) %>%
                      dplyr::distinct() %>%
                      dplyr::group_by(time_point, subject_id) %>%
                      dplyr::summarise(peptide_number=n())

my_comparisons <- list( c("baseline", "6months"), c("6months", "12months"), c("baseline", "12months") )

boxplot_before_multimatch <- ggpubr::ggboxplot(px_tp_sbj, x="time_point", y="peptide_number", color="time_point", add = "jitter", legend="none") +
                        scale_x_discrete(breaks=c("baseline","6months","12months"),
                      labels=c("B", "6M", "12M")) +
                      xlab("") + ylab("Peptide number") +
                      stat_compare_means(comparisons = my_comparisons, paired=T, method = "wilcox", label="p.signif") +
                      stat_compare_means(paired=T, label="p.format", label.y = 3000) +
        ggtitle("Before filtering")
boxplot_before_multimatch
ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_before_multimatch_filtering_boxplot.pdf"), width = 8, height = 7, units = "cm")
ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_before_multimatch_filtering_boxplot.png"), width = 8, height = 7, units = "cm")
```

```{r seq_regions_plot_unique, fig.width=3, fig.asp=1, echo=FALSE}
px_peptides_region <- px %>% dplyr::select(peptide_id, time_point, subject_id, region) %>%
                      dplyr::distinct()

ggplot(px_peptides_region, aes(x = region)) +
  geom_bar() +
  xlab("") +
  ggtitle("Before filtering")
ggsave(filename = paste0(outdir,"peptide_regions_distribution_before_multimatch_filtering.pdf"), width = 10, height = 6, units = "cm")
ggsave(filename = paste0(outdir,"peptide_regions_distribution_before_multimatch_filtering.png"), width = 10, height = 6, units = "cm")
```

# Multimatch filtering

Filtering peptides that have a multimatch (matching to more than one BCR transcriptome sequence). If those peptides are matching more than one BCR sequence, if those sequences are from different BCR clones, then these are eliminated.

```{r filter_multimatches, include=FALSE}
df_matches <- data.frame()

for (pat in c("CLAD1", "CLAD2", "CLAD3", "CLAD5", "CLAD6", "CLAD7", "CLAD8")) {
  px_match <- px %>% filter( subject_id == pat)
  px_1match <- px_match %>% filter( match_number == 1)

  df_matches <- rbind(df_matches,px_1match)
  px_gt1match <- px_match %>% filter(match_number > 1)

  tx_pat <- tx %>% filter( source == pat)

  # Checking if multi-matches belong to different sequences in the same clone and otherwise removing them
  for (p in unique(px_gt1match$peptide_id)) {
    px_p <- px_gt1match %>% filter( peptide_id == p)
    tx_p <- tx_pat %>% filter(sequence_id %in% px_p$protein_id)
    if (length(unique(tx_p$clone_id)) > 1){
      px_gt1match <- px_gt1match[!(which(px_gt1match$peptide_id == p)),]
    }
  }
  df_matches <- rbind(df_matches, px_gt1match)
}
```


Number of peptide matches per patient and time point after filtering:

```{r seq_numbers_plot_unique_after, fig.width=10, fig.asp=0.3, echo=FALSE}
ggplot(df_matches, aes(x=subject_id, fill=time_point)) +
  geom_bar(position = position_dodge()) +
  ggtitle("After filtering")
ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_after_multimatch_filtering.pdf"), width = 15, height = 6, units = "cm")
ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_after_multimatch_filtering.png"), width = 15, height = 6, units = "cm")
```

```{r number_of_peptides_after_filtering_boxplot, fig.width=3, fig.asp=1, echo=FALSE}
df_matches_tp_sbj <- df_matches %>% dplyr::select(peptide_id, time_point, subject_id) %>%
                      dplyr::distinct() %>%
                      dplyr::group_by(time_point, subject_id) %>%
                      dplyr::summarise(peptide_number=n())

my_comparisons <- list( c("baseline", "6months"), c("6months", "12months"), c("baseline", "12months") )

boxplot_after_multimatch <- ggpubr::ggboxplot(df_matches_tp_sbj, x="time_point", y="peptide_number", color="time_point", add = "jitter", legend="none") +
                        scale_x_discrete(breaks=c("baseline","6months","12months"),
                      labels=c("B", "6M", "12M")) +
                      xlab("") + ylab("Peptide number") +
                      stat_compare_means(comparisons = my_comparisons, paired=T, method = "wilcox", label="p.signif") +
                      stat_compare_means(paired=T, label="p.format", label.y = 1400) +
  ggtitle("After multimatch filtering")
boxplot_after_multimatch
ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_after_multimatch_filtering_boxplot.pdf"), width = 8, height = 7, units = "cm")
ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_after_multimatch_filtering_boxplot.png"), width = 8, height = 7, units = "cm")
```


Number of protein matches per peptide after filtering:

```{r prot_match_per_peptide_after, fig.width=10, fig.asp=0.3, echo=FALSE}
ggplot(df_matches, aes(match_number)) +
  geom_histogram(binwidth = 1) +
  scale_x_continuous(limits = c(0,10), breaks=c(1,5,10)) +
  xlab("Number of protein matches") + ylab("Peptide number")
ggsave(file = paste0(outdir,"peptide_match_distribution_after_filtering.png"), width = 8, height = 6, units = "cm")
ggsave(file = paste0(outdir,"peptide_match_distribution_after_filtering.pdf"), width = 8, height = 6, units = "cm")

```

```{r matching_regions_after_filtering, fig.width=3, fig.asp=1, echo=FALSE}
ggplot(df_matches, aes(x = region)) +
  geom_bar() +
  xlab("") +
  ggtitle("After filtering")
ggsave(filename = paste0(outdir,"peptide_regions_distribution_after_multimatch_filtering.pdf"), width = 10, height = 6, units = "cm")
ggsave(filename = paste0(outdir,"peptide_regions_distribution_after_multimatch_filtering.png"), width = 10, height = 6, units = "cm")
```

Number of amino acids overlaping in each of the cdr regions:

```{r matching_regions_after_filtering_naa, fig.width=3, fig.asp=0.8, echo=FALSE}
df_matches_cdr <- df_matches %>% filter(region != "other")
ggplot(df_matches_cdr, aes(x=region, y=cdr_aa)) +
  geom_boxplot() +
  #geom_jitter(size=1) +
  xlab("")
ggsave(filename = paste0(outdir,"peptide_regions_number_aa_distribution_after_filtering.pdf"), width = 6, height = 5, units = "cm")
ggsave(filename = paste0(outdir,"peptide_regions_number_aa_distribution_after_filtering.png"), width = 6, height = 5, units = "cm")
```


```{r matching_regions_after_filtering_jitter, fig.width=3, fig.asp=0.8, echo=FALSE}
ggplot(df_matches_cdr, aes(x=region, y=cdr_aa)) +
  geom_boxplot() +
  geom_jitter(size=0.5) +
  xlab("")
ggsave(filename = paste0(outdir,"peptide_regions_number_aa_distribution_withjitter_after_filtering.pdf"), width = 6, height = 5, units = "cm")
ggsave(filename = paste0(outdir,"peptide_regions_number_aa_distribution_withjitter_after_filtering.png"), width = 6, height = 5, units = "cm")

write.table(df_matches, file = paste0(outdir,"all_patients_peptide_protein_match_region_filtered_multimatch.tsv"), sep = "\t", quote = F, row.names = F)
```

Number of peptide matches by B-cell clone

```{r number_matches_by_clone, fig.width=10, fig.asp=0.5, echo=FALSE}
# Merging clone information
tx$protein_id_vdj = paste0(tx$sequence_id,tx$v_call,tx$d_call,tx$j_call)
df_matches_merged = merge(x=df_matches, y=tx, by.x = "protein_id_vdj", by.y = "protein_id_vdj", all.x = T)

write.table(df_matches_merged, file = paste0(outdir,"all_patients_peptide_protein_match_region_filtered_multimatch_withtxinfo.tsv"), sep = "\t", quote = F, row.names = F)


# Number of clone matches after filtering per patient and time point

df_matches_byclone <- df_matches_merged %>%
                    dplyr::select(clone_id, subject_id, time_point, v_call, d_call, j_call) %>%
                    dplyr::distinct()

ggplot(df_matches_byclone, aes(x=subject_id, fill=time_point)) +
  geom_bar(position = position_dodge()) +
  ggtitle("After filtering - number clones with peptide matches") +
  ylab("Number of clones")
ggsave(filename = paste0(outdir,"transcriptome_number_of_clones_with_peptide_matches_after_multimatch_filtering.pdf"), width = 15, height = 6, units = "cm")
ggsave(filename = paste0(outdir,"transcriptome_number_of_clones_with_peptide_matches_after_multimatch_filtering.png"), width = 15, height = 6, units = "cm")

write.table(df_matches_byclone, file = paste0(outdir,"all_patients_peptide_protein_match_byclone_region_filtered_multimatch.tsv"), sep = "\t", quote = F, row.names = F)
```

# Further filtering of peptides that have 3 aa in a cdr region

```{r filtering_by_cdr_aa, fig.width=3, fig.asp=0.8, echo=FALSE}
# FURTHER FILTERING BY CDR > 3

df_matches_cdr = df_matches %>% filter(cdr_aa >= 3)

ggplot(df_matches_cdr, aes(x=region, y=cdr_aa)) +
  geom_boxplot() +
  #geom_jitter(size=1) +
  xlab("")
ggsave(filename = paste0(outdir,"peptide_regions_number_aa_distribution_after_filtering_onlycdr.pdf"), width = 6, height = 5, units = "cm")
ggsave(filename = paste0(outdir,"peptide_regions_number_aa_distribution_after_filtering_onlycdr.png"), width = 6, height = 5, units = "cm")

write.table(df_matches_cdr, file = paste0(outdir,"all_patients_peptide_protein_match_region_filtered_multimatch_onlycdr.tsv"), sep = "\t", quote = F, row.names = F)
```

```{r filtering_by_cdr_aa_jitter, fig.width=3, fig.asp=0.8, echo=FALSE}
ggplot(df_matches_cdr, aes(x=region, y=cdr_aa)) +
  geom_boxplot() +
  geom_jitter(size=0.5) +
  xlab("")
ggsave(filename = paste0(outdir,"peptide_regions_number_aa_distribution_withjitter_after_filtering_onlycdr.pdf"), width = 6, height = 5, units = "cm")
ggsave(filename = paste0(outdir,"peptide_regions_number_aa_distribution_withjitter_after_filtering_onlycdr.png"), width = 6, height = 5, units = "cm")
```

```{r filtering_by_cdr_aa_region, fig.width=3, fig.asp=0.8, echo=FALSE}
ggplot(df_matches_cdr, aes(x = region)) +
  geom_bar() +
  xlab("") +
  ggtitle("After filtering for cdr")
ggsave(filename = paste0(outdir,"peptide_regions_distribution_after_multimatch_filtering_onlycdr.pdf"), width = 10, height = 6, units = "cm")
ggsave(filename = paste0(outdir,"peptide_regions_distribution_after_multimatch_filtering_onlycdr.png"), width = 10, height = 6, units = "cm")
```

Number of peptide matches after filtering per patient and time point, cdr only:

```{r filtering_by_cdr_aa_barplot, fig.width=10, fig.asp=0.5, echo=FALSE}
ggplot(df_matches_cdr, aes(x=subject_id, fill=time_point)) +
  geom_bar(position = position_dodge()) +
  ggtitle("After filtering only cdr")
ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_after_multimatch_filtering_onlycdr.pdf"), width = 15, height = 6, units = "cm")
ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_after_multimatch_filtering_onlycdr.png"), width = 15, height = 6, units = "cm")
```

Number of peptide matches after multimatch filtering, cdr only:

```{r number_of_peptides_after_filtering_cdr_boxplot, fig.width=3, fig.asp=1, echo=FALSE}
df_matches_cdr_tp_sbj <- df_matches_cdr %>%
                      dplyr::select(peptide_id, time_point, subject_id) %>%
                      dplyr::distinct() %>%
                      dplyr::group_by(time_point, subject_id) %>%
                      dplyr::summarise(peptide_number=n())

my_comparisons <- list( c("baseline", "6months"), c("6months", "12months"), c("baseline", "12months") )

boxplot_after_multimatch_only_cdr <- ggpubr::ggboxplot(df_matches_cdr_tp_sbj, x="time_point", y="peptide_number", color="time_point", add = "jitter", legend="none") +
                        scale_x_discrete(breaks=c("baseline","6months","12months"),
                      labels=c("B", "6M", "12M")) +
                      xlab("") + ylab("Peptide number") +
                      stat_compare_means(comparisons = my_comparisons, paired=T, method = "wilcox", label="p.signif") +
                      stat_compare_means(paired=T, label="p.format", label.y = 500)
boxplot_after_multimatch_only_cdr
ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_after_multimatch_filtering_only_cdr_boxplot.pdf"), width = 8, height = 7, units = "cm")
ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_after_multimatch_filtering_only_cdr_boxplot.png"), width = 8, height = 7, units = "cm")

boxplot_after_multimatch_only_cdr_notests <- ggpubr::ggboxplot(df_matches_cdr_tp_sbj, x="time_point", y="peptide_number", color="time_point", add = "jitter", legend="none") +
                        scale_x_discrete(breaks=c("baseline","6months","12months"),
                      labels=c("B", "6M", "12M")) +
                      xlab("") + ylab("Peptide number") +
                      stat_compare_means(paired=T, label="p.format", label.y = 350)
boxplot_after_multimatch_only_cdr_notests
ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_after_multimatch_filtering_only_cdr_boxplot_notests.pdf"), width = 8, height = 7, units = "cm")
ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_after_multimatch_filtering_only_cdr_boxplot_notests.png"), width = 8, height = 7, units = "cm")
```

Integrating transcriptomics time point information:

```{r save_tables_without_matches, include=FALSE}
# Merging tx information
df_matches_merged_cdr = df_matches_merged %>%
                          dplyr::select(peptide_id,protein_id,protein_id_vdj,subject_id,
                                        time_point,extract_time,c_primer,clone_id,
                                        v_call,d_call,j_call,cdr_aa,population) %>%
                          dplyr::filter(cdr_aa >= 3)

write.table(df_matches_merged_cdr, file = paste0(outdir,"all_patients_peptide_protein_match_region_filtered_multimatch_onlycdr_withtxinfo.tsv"),
            sep = "\t", quote = F, row.names = F)
```


Number of sequence matches in proteome, colored by which time point those sequences were matching on the transcriptome

```{r matches_transcriptome_time_point, fig.width=10, fig.asp=0.5, echo=FALSE}

ggplot(df_matches_merged_cdr, aes(x=time_point, fill=extract_time)) +
  geom_bar() +
  facet_wrap(~subject_id, scales = "free_x", ncol = 4)
ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_after_multimatch_filtering_onlycdr_colored_tp_transcriptome.pdf"), width = 22, height = 10, units = "cm")
ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_after_multimatch_filtering_onlycdr_colored_tp_transcriptome.png"), width = 22, height = 10, units = "cm")
```

```{r matches_transcriptome_time_point_boxplot, fig.width=5, fig.asp=0.6, echo=FALSE}

df_matches_merged_cdr_boxplot = df_matches_merged_cdr %>%
                                dplyr::select(peptide_id,subject_id,
                                        time_point,extract_time) %>%
                                dplyr::distinct() %>%
                                dplyr::group_by(subject_id,time_point,extract_time) %>%
                                dplyr::summarise(peptide_n=n())

df_matches_merged_cdr_boxplot$extract_time <- factor(df_matches_merged_cdr_boxplot$extract_time, levels=c("baseline","6months","12months"))

prot_boxplot_extracttime_stats <- ggpubr::ggboxplot(df_matches_merged_cdr_boxplot, x="extract_time", y="peptide_n",
                                      color="time_point", add="jitter") +
  stat_compare_means(aes(group=time_point), label="p.format", paired = T)

prot_boxplot_extracttime_stats

prot_boxplot_extracttime <- ggplot(df_matches_merged_cdr_boxplot, aes(x=extract_time, y=peptide_n, color=time_point)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(position=position_jitterdodge()) +
  xlab("Time point transcriptome") +
  #scale_color_brewer(palette=3, type="qual" ) +
  ylab("Peptide number")
prot_boxplot_extracttime

ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_after_multimatch_filtering_onlycdr_colored_tp_transcriptome_boxplot.pdf"), width = 6, height = 7, units = "cm")
ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_after_multimatch_filtering_onlycdr_colored_tp_transcriptome_boxplot.png"), width = 6, height = 7, units = "cm")
```

Number of sequence matches in proteome, colored by Ig isotype of the BCR sequence in transcriptome

```{r matches_transcriptome_ig, fig.width=10, fig.asp=0.5, echo=FALSE}
ggplot(df_matches_merged_cdr, aes(x=time_point, fill=c_primer)) +
  geom_bar() +
  facet_wrap(~subject_id, scales = "free_x", ncol = 4)
ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_after_multimatch_filtering_onlycdr_colored_by_ig.pdf"), width = 22, height = 10, units = "cm")
ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_after_multimatch_filtering_onlycdr_colored_by_ig.png"), width = 22, height = 10, units = "cm")
```

```{r matches_transcriptome_ig_boxplot, fig.width=5, fig.asp=0.6, echo=FALSE}

df_matches_merged_cdr_boxplot = df_matches_merged_cdr %>%
                                dplyr::select(peptide_id,subject_id,
                                        time_point,c_primer) %>%
                                dplyr::distinct() %>%
                                dplyr::group_by(subject_id,time_point,c_primer) %>%
                                dplyr::summarise(peptide_n=n())


prot_boxplot_ig_stats <- ggpubr::ggboxplot(df_matches_merged_cdr_boxplot, x="c_primer", y="peptide_n",
                                      color="time_point", add="jitter") +
   scale_x_discrete(breaks=c("baseline","6months","12months"),
                      labels=c("B", "6M", "12M")) +
  stat_compare_means(aes(group=time_point), label="p.format", paired = T) +
  #xlab("") +
  #scale_color_brewer(palette=3, type="qual" ) +
  ylab("Peptide number") +
  theme(axis.ticks.x = element_line(), axis.text.x=element_text(family = "arial"))
prot_boxplot_ig_stats

prot_boxplot_ig <- ggplot(df_matches_merged_cdr_boxplot, aes(x=c_primer, y=peptide_n, color=time_point)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(position=position_jitterdodge()) +
  #xlab("") +
  #scale_color_brewer(palette=3, type="qual" ) +
  ylab("Peptide number")
prot_boxplot_ig

ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_after_multimatch_filtering_onlycdr_colored_ig_boxplot.pdf"), width = 6, height = 7, units = "cm")
ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_after_multimatch_filtering_onlycdr_colored_ig_boxplot.png"), width = 6, height = 7, units = "cm")
```

Number of sequence matches in proteome, colored by population of origin of the BCR sequence in transcriptome

```{r matches_transcriptome_pop, fig.width=10, fig.asp=0.5, echo=FALSE}
ggplot(df_matches_merged_cdr, aes(x=time_point, fill=population)) +
  geom_bar() +
  facet_wrap(~subject_id, scales = "free_x", ncol = 4)
ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_after_multimatch_filtering_onlycdr_colored_by_population.pdf"), width = 22, height = 10, units = "cm")
ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_after_multimatch_filtering_onlycdr_colored_by_population.png"), width = 22, height = 10, units = "cm")
```

```{r matches_transcriptome_pop_boxplot, fig.width=5, fig.asp=0.6, echo=FALSE}

df_matches_merged_cdr_boxplot = df_matches_merged_cdr %>%
                                dplyr::select(peptide_id,subject_id,
                                        time_point,population) %>%
                                dplyr::distinct() %>%
                                dplyr::group_by(subject_id,time_point,population) %>%
                                dplyr::summarise(peptide_n=n())

df_matches_merged_cdr_boxplot$population <- factor(df_matches_merged_cdr_boxplot$population, levels=c("N","M","DN","P"))

prot_boxplot_pop_stats <- ggpubr::ggboxplot(df_matches_merged_cdr_boxplot, x="population", y="peptide_n",
                                      color="time_point", add="jitter") +
   scale_x_discrete(breaks=c("baseline","6months","12months"),
                      labels=c("B", "6M", "12M")) +
  stat_compare_means(aes(group=time_point), label="p.format", paired = T) +
  #xlab("") +
  #scale_color_brewer(palette=3, type="qual" ) +
  ylab("Peptide number") +
  theme(axis.ticks.x = element_line(), axis.text.x=element_text(family = "arial"))
prot_boxplot_pop_stats

prot_boxplot_pop <- ggplot(df_matches_merged_cdr_boxplot, aes(x=population, y=peptide_n, color=time_point)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(position=position_jitterdodge()) +
  #xlab("") +
  #scale_color_brewer(palette=3, type="qual" ) +
  ylab("Peptide number")
prot_boxplot_pop

ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_after_multimatch_filtering_onlycdr_colored_population_boxplot.pdf"), width = 6, height = 7, units = "cm")
ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_after_multimatch_filtering_onlycdr_colored_population_boxplot.png"), width = 6, height = 7, units = "cm")
```

## Analysis of number of clones
Number of clone matches after filtering per patient and time point, cdr only

```{r matches_transcriptome_clones, fig.width=10, fig.asp=0.5, echo=FALSE}
df_matches_byclone_cdr = df_matches_merged_cdr[,c("clone_id", "subject_id", "time_point", "v_call", "d_call", "j_call")] %>% distinct()

ggplot(df_matches_byclone_cdr, aes(x=subject_id, fill=time_point)) +
  geom_bar(position = position_dodge()) +
  ggtitle("After filtering only cdr3") +
  xlab("") + ylab("Number of clones")
ggsave(filename = paste0(outdir,"transcriptome_clone_proteome_matches_after_multimatch_filtering_onlycdr.pdf"), width = 15, height = 6, units = "cm")
ggsave(filename = paste0(outdir,"transcriptome_clone_proteome_matches_after_multimatch_filtering_onlycdr.png"), width = 15, height = 6, units = "cm")

write.table(df_matches_byclone_cdr, file = paste0(outdir,"all_patients_peptide_protein_match_byclone_region_filtered_multimatch_onlycdr.tsv"), sep = "\t", quote = F, row.names = F)
```

Figure with all boxplots:

```{r matches_transcriptome_figure, fig.width=8, fig.asp=0.8, echo=FALSE}

ggarrange(boxplot_after_multimatch_only_cdr_notests, prot_boxplot_extracttime, prot_boxplot_ig, prot_boxplot_pop, ncol = 2, nrow=2, labels = "AUTO" )

ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_after_multimatch_filtering_onlycdr_figure.pdf"), width = 18, height = 12, units = "cm")
ggsave(filename = paste0(outdir,"transcriptome_proteome_matches_after_multimatch_filtering_onlycdr_figure.png"), width = 18, height = 12, units = "cm")
```

# Upset plots per patient

```{r upset_proteomics, fig.width=10, fig.asp=0.8, echo=FALSE}
for (pat in c("CLAD1", "CLAD2", "CLAD3", "CLAD5", "CLAD6", "CLAD7", "CLAD8")) {

  df_matches_merged_pat <- df_matches_merged %>% filter(subject_id == pat)
  px_baseline <- df_matches_merged_pat %>% filter(time_point == "baseline")
  px_6months <- df_matches_merged_pat %>% filter(time_point == "6months")
  px_12months <- df_matches_merged_pat %>% filter(time_point == "12months")

  tx_pat <- tx %>% filter(source == pat)
  tx_baseline <- tx_pat %>% filter(extract_time == "baseline")
  tx_6months <- tx_pat %>% filter(extract_time == "6months")
  tx_12months <- tx_pat %>% filter(extract_time == "12months")


  # Sequence  overlap
  listinput <- list(px_baseline$protein_id_vdj,px_6months$protein_id_vdj,px_12months$protein_id_vdj)
  names(listinput) <- c("px_baseline","px_6months","px_12months")

  svg(filename = paste0(outdir,"transcriptome_proteome_sequence_overlap_", pat, ".svg"),
      width = 20*0.4, height=10*0.4)

  print(upset( fromList(listinput),
               nsets = length(listinput),
               nintersects = NA,
               keep.order=TRUE,
               order.by = "freq",
               #group.by = "sets",
               point.size = 3.5,
               line.size=2,
               mainbar.y.label = "Sequence intersections",
               sets.x.label = "Sequence per time point"))

  dev.off()

  png(filename = paste0(outdir,"transcriptome_proteome_sequence_overlap_", pat, ".png"),
      res=600,width = 15, height=10, units="cm")
  print(upset( fromList(listinput),
               nsets = length(listinput),
               nintersects = NA,
               keep.order=TRUE,
               order.by = "freq",
               #group.by = "sets",
               point.size = 3.5,
               line.size=2,
               mainbar.y.label = "Sequence intersections",
               sets.x.label = "Sequence per time point"))

  dev.off()

  # Clone overlap
  listinput <- list(px_baseline$clone_id,px_6months$clone_id,px_12months$clone_id)
  names(listinput) <- c("px_baseline","px_6months","px_12months")


  svg(filename = paste0(outdir,"transcriptome_proteome_clone_overlap_", pat, ".svg"),
      width = 20*0.4, height=10*0.4)
  print(upset( fromList(listinput),
               nsets = length(listinput),
               nintersects = NA,
               keep.order=TRUE,
               order.by = "freq",
               #group.by = "sets",
               point.size = 3.5,
               line.size=2,
               mainbar.y.label = "Clone intersections",
               sets.x.label = "Clones per time point"))

  dev.off()

  png(filename = paste0(outdir,"transcriptome_proteome_clone_overlap_", pat, ".png"),
      res=600,width = 15, height=10, units="cm")
  print(upset( fromList(listinput),
               nsets = length(listinput),
               nintersects = NA,
               keep.order=TRUE,
               order.by = "freq",
               #group.by = "sets",
               point.size = 3.5,
               line.size=2,
               mainbar.y.label = "Clone intersections",
               sets.x.label = "Clones per time point"))

  dev.off()


}
```
