---
output:
  html_document:
    toc: true                               # table of contents
    toc_float: true                         # float the table of contents to the left of the main document content
    toc_depth: 3                            # header levels 1,2,3
    theme: default
    number_sections: false                  # add section numbering to headers
    df_print: paged                         # tables are printed as an html table with support for pagination over rows and columns
    css: ./nf-core_style.css
    highlight: pygments
    pdf_document: true
  html_notebook:
    toc: yes
bibliography: ./references.bibtex
---


---
title: "Repertoire analysis"
subtitle: "Clonal abundance, diversity and V-family gene usage"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: html_document
---


```{r setup, include=FALSE}
library(knitr)
library(kableExtra)
library(dplyr)
library(alakazam)
library(shazam)
library(stringr)

library(data.table)
library(igraph)
library(gplots)
library(circlize)
library(UpSetR)
library(gtools)
library(ggpubr)
library(cowplot)

theme_set(theme_bw(base_family = "ArialMT") +
            theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), text = element_text(family="ArialMT")))
knitr::opts_chunk$set(echo = FALSE)
```
***

# Bcellmagic analysis pipeline

## Pipeline overview

## Number of sequences

Number of reads for each of the samples and number of sequences left after representative analysis steps.

```{r seq_numbers, echo=FALSE, results='asis'}
tab_seqs <- read.table("Table_sequences_process.tsv", header=TRUE, sep="\t", check.names = FALSE)
kable(tab_seqs) %>%
    kable_styling("hover") %>%
    scroll_box(width = "100%", height = "400px")
```

Plotting number of unique sequences

```{r seq_numbers_plot_unique, fig.width=10, fig.asp=0.3, echo=FALSE}
tab_seqs$Extraction_time <- factor(tab_seqs$Extraction_time, levels=c("baseline","6months","12months"))
g4 <- ggplot(tab_seqs, aes(fill=Extraction_time, y=Unique, x=Population)) +
  geom_bar(position = "dodge", stat="identity") +
  xlab("") +
  ylab("Sequence number") +
  ggtitle("Unique sequence numbers per sample") +
  facet_grid(cols=vars(Source), scales = "free", drop = T) +
  theme(axis.text.x = element_text(angle=45, hjust = 1, vjust = 1))
g4
```


Plotting number of representative 2 sequences

```{r seq_numbers_plot_rep2, fig.width=10, fig.asp=0.3, echo=FALSE}
g4 <- ggplot(tab_seqs, aes(fill=Extraction_time, y=Representative_2, x=Population)) +
  geom_bar(position = "dodge", stat="identity") +
  xlab("") +
  ylab("Sequence number") +
  ggtitle("Representative2 sequence numbers per sample") +
  facet_grid(cols=vars(Source), scales = "free", drop = T) +
  theme(axis.text.x = element_text(angle=45, hjust = 1, vjust = 1))
g4
```


Plotting number of Igblast identified sequences


```{r seq_numbers_plot, fig.width=10, fig.asp=0.3, echo=FALSE}
g4 <- ggplot(tab_seqs, aes(fill=Extraction_time, y=Igblast, x=Population)) +
  geom_bar(position = "dodge", stat="identity") +
  xlab("") +
  ylab("Sequence number") +
  ggtitle("Igblast identified sequence number per sample") +
  facet_grid(cols=vars(Source), scales = "free", drop = T) +
  theme(axis.text.x = element_text(angle=45, hjust = 1, vjust = 1))
g4
```



```{r read_data, include=FALSE}
# Reading data tables and ceating output folders
#datadir <- "."
outdir <- "repertoire_comparison_tp"
# setwd to results folder (containing alakazam, shazam, etc. folders)
### Read all the tables as produced by the pipeline in the current folder and joins them together in the df_all dataframe
#all_files <- system(paste0("find '",datadir,"' -name '*germ-pass.tsv'"), intern=T)
dir.create(outdir)
diversity_dir <- paste(outdir, "Diversity", sep="/")
abundance_dir <- paste(outdir, "Abundance", sep="/")
vfamily_dir <- paste(outdir, "V_family", sep="/")
isotype_dir <- paste(outdir, "Isotype", sep="/")
dir.create(diversity_dir)
dir.create(abundance_dir)
dir.create(vfamily_dir)
dir.create(isotype_dir)
# Generate one big dataframe from all patient dataframes
#df_all = data.frame()
#for (file in all_files){
#    fname = file
#    print(fname)
#    df_pat <- read.csv(fname, sep="\t")
#    df_all <- rbind(df_all, df_pat)
#}
#write.table(df_all, paste0(outdir,"/all_data.tsv"), sep = "\t", quote=F, row.names = F, col.names = T)
df_all <- read.csv("./repertoire_comparison_tp/all_data.tsv", sep="\t")

# Remove underscores in these columns
df_all$treatment <- sapply(df_all$treatment, function(x) str_replace(as.character(x), "_", ""))
df_all$source <- sapply(df_all$source, function(x) str_replace(as.character(x), "_", ""))
df_all$extract_time <- sapply(df_all$extract_time, function(x) str_replace(as.character(x), "_", ""))
df_all$population <- sapply(df_all$population, function(x) str_replace(as.character(x), "_", ""))

df_all$source <- sapply(df_all$source, function(x) substring(as.character(x),1,5))

# Annotate sample and samplepop (sample + population) by add ing all the conditions
df_all$sample <- as.factor(paste(df_all$treatment, df_all$extract_time, df_all$source, sep="_"))
df_all$sample_pop <- as.factor(paste(df_all$treatment, df_all$extract_time, df_all$source, df_all$population, sep="_"))
# Set number of bootrstraps
nboot = 200

# Removing Cladribin_6months_CLAD4_N because of too little sequences
df_all <- df_all[!(df_all$sample_pop=="Cladribin_baseline_CLAD4_N"),]

```

# Clonal abundance

For plotting the clonal abundance, the clones were ordered by size from bigger clones to smaller clones (x-axis, Rank).
The Abundance of each clone was represented as the percentage of unique sequences in the clone, with respect to the total
number of unique sequences in that subject (By Patient) or in the B-cell or T-cell sample (By Cell Population).

To correct for the different number of sequences in each of the samples, the Bootstrapping technique was employed,
in which `r nboot` random bootstrap samples were taken, with size the number of sequences in the sample with less sequences (N).
The solid line shows the mean Abundance of the bootstrap samples, whereas the transparent area shows the full Abundance
range of the bootstrap samples.

All clonal abundance plots and tables with abundance values can be found under `repertoire_analysis/Abundance`.

## Clonal abundance per subject
```{r clonal_abundance, echo=FALSE, include=FALSE}
abund <- estimateAbundance(df_all, group = "sample", ci=0.95, nboot=nboot)
abund@abundance$treatment <- sapply(abund@abundance$sample, function(x) unlist(strsplit(as.character(x), "_"))[1])
abund@abundance$time_point <- sapply(abund@abundance$sample, function(x) unlist(strsplit(as.character(x), "_"))[2])
abund@abundance$patient <- sapply(abund@abundance$sample, function(x) unlist(strsplit(as.character(x), "_"))[3])
abund_main <- paste0("Clonal abundance (N=", abund@n[1], ")")
```
```{r clonal_abundance2, fig.height=3, fig.width=12, echo=FALSE}
p_ca <- ggplot(abund@abundance, aes(x = rank, y = p,
                                    group = sample)) +
    geom_ribbon(aes(ymin = lower,
                    ymax = upper, fill = time_point), alpha = 0.4) +
    geom_line(aes(color = time_point)) +
    ggtitle(abund_main) +
    xlab("Rank") + ylab("Abundance") +
    scale_x_log10(limits = NULL,
                breaks = scales::trans_breaks("log10", function(x) 10^x),
                labels = scales::trans_format("log10", scales::math_format(10^.x))) +
    scale_y_continuous(labels = scales::percent) +
    facet_grid(cols = vars(patient), rows = vars(treatment), scales="free", drop = T)
p_ca
```

```{r include = FALSE}
ggsave(plot=p_ca, filename = paste0(abundance_dir,"/Clonal_abundance_subject.pdf"), device="pdf", width = 25, height = 6, units="cm")
ggsave(plot=p_ca, filename = paste0(abundance_dir,"/Clonal_abundance_subject.png"), device="png", width = 25, height = 6, units="cm")
write.table(abund@abundance, file = paste0(abundance_dir, "/Clonal_abundance_data_subject.tsv"), sep="\t", quote = F, row.names = F)
```

## Count clones per subject

```{r count_clones}
df_countclones_copy <- countClones(df_all,groups="sample",copy = "duplicate_count")
df_countclones_copy$treatment <- sapply(df_countclones_copy$sample, function(x) unlist(strsplit(as.character(x), "_"))[1])
df_countclones_copy$time_point <- sapply(df_countclones_copy$sample, function(x) unlist(strsplit(as.character(x), "_"))[2])
df_countclones_copy$patient <- sapply(df_countclones_copy$sample, function(x) unlist(strsplit(as.character(x), "_"))[3])
write.table(df_countclones_copy, file = paste0(abundance_dir, "/Clonal_numbers_data_subject.tsv"), sep = "\t", quote = F, row.names = F)
```

## Clonal abundance per cell population

If different types of B-cell or T-cell populations are provided, here
the clonal abundance is plotted for each patient and B / T-cell population.

```{r clonal_abundance_pop, echo=FALSE, include=FALSE}

abund_pop <- estimateAbundance(df_all, clone="clone_id", group = "sample_pop", ci=0.95, nboot=nboot)
abund_pop@abundance$treatment <- sapply(abund_pop@abundance$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[1])
abund_pop@abundance$time_point <- sapply(abund_pop@abundance$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[2])
abund_pop@abundance$patient <- sapply(abund_pop@abundance$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[3])
abund_pop@abundance$population <- sapply(abund_pop@abundance$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[4])
abund_main <- paste0("Clonal abundance (N=", abund_pop@n[1], ")")
```

```{r clonal_abundance_pop2, fig.width=12, echo=FALSE}
abund_pop@abundance$population <- factor(abund_pop@abundance$population, levels = c("N", "M", "DN", "P"))
abund_pop@abundance$time_point <- factor(abund_pop@abundance$time_point, levels = c("baseline", "6months", "12months"))
pop_ca <- ggplot(abund_pop@abundance, aes(x = rank, y = p,
                                group = sample_pop)) +
    geom_ribbon(aes(ymin = lower,
                    ymax = upper, fill = time_point), alpha = 0.4) +
    geom_line(aes(color = time_point)) +
    ggtitle(abund_main) +
    xlab("Rank") + ylab("Abundance") +
    labs(color="Time Point", fill="Time Point") +
    scale_x_log10(limits = NULL,
                    breaks = scales::trans_breaks("log10", function(x) 10^x),
                    labels = scales::trans_format("log10", scales::math_format(10^.x))) +
    scale_y_continuous(labels = scales::percent) +
    facet_grid(cols=vars(patient), rows=vars(population), scales="free", drop = T)
pop_ca
```
```{r include=FALSE}
ggsave(plot=pop_ca, filename = paste0(abundance_dir,"/Clonal_abundance_patient_population.pdf"), device="pdf",
        width = 30, height = 20, units="cm")
ggsave(plot=pop_ca, filename = paste0(abundance_dir,"/Clonal_abundance_patient_population.png"), device="png",
        width = 30, height = 20, units="cm")
write.table(abund_pop@abundance, file = paste0(abundance_dir, "/Clonal_abundance_data_cell_population.tsv"), sep="\t", quote = F, row.names = F)

```

## Count clones per population

```{r count_clones_pop_2, echo=F}
count_clones_pop <- countClones(df_all, groups = "sample_pop", copy = "duplicate_count")

count_clones_pop$treatment <- sapply(count_clones_pop$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[1])
count_clones_pop$time_point <- sapply(count_clones_pop$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[2])
count_clones_pop$patient <- sapply(count_clones_pop$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[3])
write.table(count_clones_pop, file = paste0(abundance_dir, "/Clonal_numbers_data_subject_population.tsv"), sep = "\t", quote = F, row.names = F)

```

## Clonal diversification index

```{r clonal_diversification, echo=FALSE, include=FALSE}

# Calculate gini index
clonal_diversification_index <- function(df){
  # Computing cumulative frequence distribution
  clone_sequences <- df$seq_count
  R_vec <- c()
  n = 0
  for (j in clone_sequences){
    n = n + j
    R_vec <- append(R_vec,n)
  }
  R_vec_norm <- R_vec / n

  m = 0
  g = 0
  for (i in c(1:length(R_vec_norm))){
    g = g + (R_vec_norm[i] - i/length(R_vec_norm))
  }
  gini_index = g/length(R_vec_norm)
  return(gini_index)
}

clonal_diversification_mean <- c()
clonal_diversification_std <- c()
sample_name <- c()
clones_sample_pop <- split(count_clones_pop, count_clones_pop$sample_pop)
# Removing CLAD4  N
clones_sample_pop$Cladribin_baseline_CLAD4_N <- NULL
clones_sample_pop$Cladribin_12months_CLAD4_N <- NULL
clones_sample_pop$Cladribin_6months_CLAD4_N <- NULL
cloneN <- count(count_clones_pop, vars="sample_pop")
sample_size <- min(cloneN$n)

# Calculate gini index for bootstrap samples
for (i in c(1:length(clones_sample_pop))) {
  gi_boot <- c()
  for (r in c(1:nboot)) {
    # sample N clones according to sample size
    sample_df <- sample_n(clones_sample_pop[[i]], size = sample_size, replace = F, weight = clones_sample_pop[[i]]$seq_freq)
    gi <- clonal_diversification_index(sample_df)
    gi_boot <- append(gi_boot, gi)
  }
  clonal_diversification_mean <- append(clonal_diversification_mean, mean(gi_boot))
  clonal_diversification_std <- append(clonal_diversification_std, sd(gi_boot))
  sample_name <- append(sample_name, names(clones_sample_pop)[i])
}

df_gini_index <- data.frame(sample_pop = sample_name, clonal_diversification_mean = clonal_diversification_mean, clonal_diversification_std = clonal_diversification_std)
df_gini_index$treatment <- sapply(df_gini_index$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[1])
df_gini_index$time_point <- sapply(df_gini_index$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[2])
df_gini_index$patient <- sapply(df_gini_index$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[3])
df_gini_index$population <- sapply(df_gini_index$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[4])
df_gini_index$time_point <- factor(df_gini_index$time_point, c("baseline","6months","12months"))
df_gini_index$population <- factor(df_gini_index$population, c("N", "M", "DN", "P"))

write.table(df_gini_index, paste0(abundance_dir, "/Clonal_diversification_index_table.tsv"),
            sep="\t", quote = F, row.names = F)

```

```{r clonal_diversification_boxplot, fig.width=6, fig.asp=0.5, echo=FALSE}
# boxplot_gini <- ggplot(df_gini_index, aes(y=clonal_diversification_mean, x=time_point)) +
#     geom_boxplot() +
#     geom_point(position=position_dodge(), stat="identity", size = 1) +
#     #geom_errorbar(aes(ymin=d-d_sd, ymax=d+d_sd), width = .2, position=dodge) +
#     xlab("") + ylab("Clonal diversification") +
#     ggtitle(paste0("Clonal diversification index (N = ",sample_size,")")) +
#     facet_grid(cols=vars(population), drop=T, scales = "free") +
#     theme(axis.text.x = element_text(angle = 45, hjust = 1))
# boxplot_gini

# Boxplot with Wilcoxon paired tests (ggpubr)
my_comparisons <- list( c("baseline", "6months"), c("6months", "12months"), c("baseline", "12months") )

boxplot_gini <- ggboxplot(df_gini_index, x = "time_point", y = "clonal_diversification_mean",
        add = "jitter", legend = "none",
         color = "time_point", line.color = "gray", line.size = 0.4,
         facet.by = "population", ncol = 4, short.panel.labs = T) +
        scale_x_discrete(breaks=c("baseline","6months","12months"),
                      labels=c("B", "6M", "12M")) +
        #theme(axis.text.x = element_text(angle = 45, hjust=1), legend.position = "none") +
        xlab("") + ylab("Clonal diversification") +
        stat_compare_means(comparisons = my_comparisons, paired =T, method = "wilcox", label="p.signif") +
        stat_compare_means(label="p.format", label.y=0.6)
boxplot_gini

boxplot_gini_nokruskal <- ggboxplot(df_gini_index, x = "time_point", y = "clonal_diversification_mean",
        add = "jitter", legend = "none",
         color = "time_point", line.color = "gray", line.size = 0.4,
         facet.by = "population", ncol = 4, short.panel.labs = T) +
        scale_x_discrete(breaks=c("baseline","6months","12months"),
                      labels=c("B", "6M", "12M")) +
        #theme(axis.text.x = element_text(angle = 45, hjust=1), legend.position = "none") +
        xlab("") + ylab("Clonal diversification") +
        stat_compare_means(comparisons = my_comparisons, paired =T, method = "wilcox", label="p.signif")
boxplot_gini_nokruskal
```

```{r, echo=F}
ggsave(plot=boxplot_gini, filename = paste0(abundance_dir,"/Clonal_diversification_index_patient_population.png"), device="png",
       width = 20, height = 7, units="cm")
ggsave(plot=boxplot_gini, filename = paste0(abundance_dir,"/Clonal_diversification_index_patient_population.pdf"), device="pdf",
       width=20, height=7, units="cm")

```




```{r clonal_sizes, echo = F}

# Calculate clonal size percentages
clone_percentage_gt_size <- function(df, size){
  # Computing cumulative frequence distribution
  clone_sequences <- df$seq_count
  df_size <- df[which(clone_sequences >= size),]
  clone_N <- nrow(df)
  clone_size_N <- nrow(df_size)
  percentage_size <- clone_size_N * 100 / clone_N
  return(percentage_size)
}

percentage_gt_2 <- c()
percentage_gt_5 <- c()
percentage_gt_10 <- c()
percentage_gt_50 <- c()
percentage_gt_100 <- c()
percentage_gt_500 <- c()
percentage_gt_1000 <- c()
percentage_gt_2 <- c()
sd_percentage_gt_2 <- c()
sd_percentage_gt_5 <- c()
sd_percentage_gt_10 <- c()
sd_percentage_gt_50 <- c()
sd_percentage_gt_100 <- c()
sd_percentage_gt_500 <- c()
sd_percentage_gt_1000 <- c()

sample_name <- c()
clones_sample_pop <- split(count_clones_pop, count_clones_pop$sample_pop)
clones_sample_pop$Cladribin_baseline_CLAD4_N <- NULL
clones_sample_pop$Cladribin_12months_CLAD4_N <- NULL
clones_sample_pop$Cladribin_6months_CLAD4_N <- NULL
cloneN <- count(count_clones_pop, vars="sample_pop")
sample_size <- min(cloneN$n)

for (i in c(1:length(clones_sample_pop))) {
  percentage_gt_2_boot <- c()
  percentage_gt_5_boot <- c()
  percentage_gt_10_boot <- c()
  percentage_gt_50_boot <- c()
  percentage_gt_100_boot <- c()
  percentage_gt_500_boot <- c()
  percentage_gt_1000_boot <- c()
  for (r in c(1:nboot)) {
    sample_df <- sample_n(clones_sample_pop[[i]], size = sample_size, replace = F, weight = clones_sample_pop[[i]]$seq_freq)
    percentage_gt_2_boot <- append(percentage_gt_2_boot, clone_percentage_gt_size(sample_df, 2))
    percentage_gt_5_boot <- append(percentage_gt_5_boot, clone_percentage_gt_size(sample_df, 5))
    percentage_gt_10_boot <- append(percentage_gt_10_boot, clone_percentage_gt_size(sample_df, 10))
    percentage_gt_50_boot <- append(percentage_gt_50_boot, clone_percentage_gt_size(sample_df, 50))
    percentage_gt_100_boot <- append(percentage_gt_100_boot, clone_percentage_gt_size(sample_df, 100))
    percentage_gt_500_boot <- append(percentage_gt_500_boot, clone_percentage_gt_size(sample_df, 500))
    percentage_gt_1000_boot <- append(percentage_gt_1000_boot,clone_percentage_gt_size(sample_df, 1000))
  }
  percentage_gt_2 <- append(percentage_gt_2, mean(percentage_gt_2_boot))
  percentage_gt_5 <- append(percentage_gt_5, mean(percentage_gt_5_boot))
  percentage_gt_10 <- append(percentage_gt_10, mean(percentage_gt_10_boot))
  percentage_gt_50 <- append(percentage_gt_50, mean(percentage_gt_50_boot))
  percentage_gt_100 <- append(percentage_gt_100, mean(percentage_gt_100_boot))
  percentage_gt_500 <- append(percentage_gt_500, mean(percentage_gt_500_boot))
  percentage_gt_1000 <- append(percentage_gt_1000, mean(percentage_gt_1000_boot))
  sd_percentage_gt_2 <- append(sd_percentage_gt_2, sd(percentage_gt_2_boot))
  sd_percentage_gt_5 <- append(sd_percentage_gt_5, sd(percentage_gt_5_boot))
  sd_percentage_gt_10 <- append(sd_percentage_gt_10, sd(percentage_gt_10_boot))
  sd_percentage_gt_50 <- append(sd_percentage_gt_50, sd(percentage_gt_50_boot))
  sd_percentage_gt_100 <- append(sd_percentage_gt_100, sd(percentage_gt_100_boot))
  sd_percentage_gt_500 <- append(sd_percentage_gt_500, sd(percentage_gt_500_boot))
  sd_percentage_gt_1000 <- append(sd_percentage_gt_1000, sd(percentage_gt_1000_boot))
  sample_name <- append(sample_name, names(clones_sample_pop)[i])
}

clones_percentages <- data.frame(sample_pop = sample_name,
                                 gt_2 = percentage_gt_2,
                                 gt_5 = percentage_gt_5,
                                 gt_10 = percentage_gt_10,
                                 gt_50 = percentage_gt_50,
                                 gt_100 = percentage_gt_100,
                                 gt_500 = percentage_gt_500,
                                 gt_1000 = percentage_gt_1000)
clones_percentages <- clones_percentages %>% distinct(sample_pop,.keep_all=T)
clones_sd_percentages <- data.frame(sample_pop = sample_name,
                                    gt_2 = sd_percentage_gt_2,
                                    gt_5 = sd_percentage_gt_5,
                                    gt_10 = sd_percentage_gt_10,
                                    gt_50 = sd_percentage_gt_50,
                                    gt_100 = sd_percentage_gt_100,
                                    gt_500 = sd_percentage_gt_500,
                                    gt_1000 = sd_percentage_gt_1000)
clones_sd_percentages <- clones_sd_percentages %>% distinct(sample_pop, .keep_all=T)
clones_percentages_melt <- melt(data.table(clones_percentages), id = "sample_pop", value.name = "percentage")
clones_sd_percentages_melt <- melt(data.table(clones_sd_percentages), id="sample_pop", value.name = "sd")

clones_percentages_df <- merge(x=clones_percentages_melt, y=clones_sd_percentages_melt, by=c("sample_pop","variable"))


clones_percentages_df$treatment <- sapply(clones_percentages_df$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[1])
clones_percentages_df$time_point <- sapply(clones_percentages_df$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[2])
clones_percentages_df$time_point <- factor(clones_percentages_df$time_point,levels=c("baseline","6months","12months"))
clones_percentages_df$patient <- sapply(clones_percentages_df$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[3])
clones_percentages_df$population <- sapply(clones_percentages_df$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[4])

clones_percentages_df <- clones_percentages_df %>% distinct()

write.table(clones_percentages_df, paste0(abundance_dir,"/Clonal_size_percentage_patient_population_table.tsv" ),
            sep="\t", quote = F, row.names = F)
```

```{r clonal_percentage_boxplot, fig.width=6, fig.asp=0.5,echo=FALSE}
clones_percentage_one_C <- clones_percentages_df[which(clones_percentages_df$variable %in% c("gt_100")),]
clones_percentage_one_C$time_point <- factor(clones_percentage_one_C$time_point,
                                             levels=c("baseline","6months","12months"))
clones_percentage_one_C$population <- factor(clones_percentage_one_C$population,
                                             levels=c("N", "M", "DN", "P"))
# boxplot_clone_size <- ggplot(clones_percentage_one_C, aes(y=percentage, x=time_point)) +
#     geom_boxplot() +
#     geom_point(position=dodge, stat="identity", size = 1) +
#     #geom_errorbar(aes(ymin=d-d_sd, ymax=d+d_sd), width = .2, position=dodge) +
#     xlab("") + ylab("Percentage") +
#     ggtitle(paste0("Clonal size percentage (N = ",sample_size,") greater than 10")) +
#     facet_grid(cols=vars(population), drop=T, scales = "free") +
#     theme(axis.text.x = element_text(angle = 45, hjust = 1))
# boxplot_clone_size

my_comparisons <- list( c("baseline", "6months"), c("6months", "12months"), c("baseline", "12months") )

boxplot_clone_size <- ggboxplot(clones_percentage_one_C,
                                x = "time_point", y = "percentage",
                                add = "jitter", legend = "none",
                                color = "time_point", line.color = "gray", line.size = 0.4,
                                facet.by = "population", ncol = 4, short.panel.labs = T) +
                      scale_x_discrete(breaks=c("baseline","6months","12months"),
                                        labels=c("B", "6M", "12M")) +
        #theme(axis.text.x = element_text(angle = 45, hjust=1), legend.position = "none") +
        xlab("") + ylab("% Clones > 100") +
        stat_compare_means(comparisons = my_comparisons, paired =T, method = "wilcox", label="p.signif") +
        stat_compare_means(label="p.format", label.y = 19)

boxplot_clone_size

boxplot_clone_size_nokruskal <- ggboxplot(clones_percentage_one_C,
                                x = "time_point", y = "percentage",
                                add = "jitter", legend = "none",
                                color = "time_point", line.color = "gray", line.size = 0.4,
                                facet.by = "population", ncol = 4, short.panel.labs = T) +
                      scale_x_discrete(breaks=c("baseline","6months","12months"),
                                        labels=c("B", "6M", "12M")) +
        #theme(axis.text.x = element_text(angle = 45, hjust=1), legend.position = "none") +
        xlab("") + ylab("% Clones > 100") +
        stat_compare_means(comparisons = my_comparisons, paired =T, method = "wilcox", label="p.signif")
boxplot_clone_size_nokruskal
```

```{r, echo=F}
ggsave(plot=boxplot_clone_size, filename = paste0(abundance_dir,"/Clones_percentages_greater_than_100.png"), device="png",
       width = 20, height = 7, units="cm")
ggsave(plot=boxplot_clone_size, filename = paste0(abundance_dir,"/Clones_percentages_greater_than_100.pdf"), device="pdf",
       width=20, height=7, units="cm")
```

## Count clones per population

```{r count_clones_pop}
df_countclones_copy <- countClones(df_all,groups="sample_pop",copy = "duplicate_count")
df_countclones_copy$treatment <- sapply(df_countclones_copy$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[1])
df_countclones_copy$time_point <- sapply(df_countclones_copy$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[2])
df_countclones_copy$patient <- sapply(df_countclones_copy$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[3])
df_countclones_copy$population <- sapply(df_countclones_copy$sample_pop, function(x)
unlist(strsplit(as.character(x), "_"))[4])
write.table(df_countclones_copy, file = paste0(abundance_dir, "/Clonal_numbers_data_subject_population.tsv"), sep = "\t", quote = F, row.names = F)
```

# Clonal diversity

The clonal diversity $D$ of the repertoire was calculated according to the general formula of Hill Diversity
numbers:

$$
\begin{aligned}
    ^{q}D = \left( \sum_{i=1}^Rp_i^q \right)^{1/(1-q)}
\end{aligned}
$$

where:

* $p_i$ is the proportion of unique sequences belonging to clone $i$.
* $q$ are the values of the different diversity numbers.
* $R$ is the Richness, the number of different clones in the sample.

At $q=1$ the function is undefined and the limit to zero equals the exponential of the Shannon Entropy:

$$
\begin{aligned}
    ^{1}D = exp \left(  \sum_{i=1}^Rp_i ln(p_i)  \right)
\end{aligned}
$$

The intuition about the different Hill Diversity values is the following:

* At $q=0$ the diversity index equals the number of clones in the sample.
* At $q=1$ the diversity index is the geometric mean of the clones in the sample,
weighted by their proportion in the sample.
* At $q>1$ more weight is given to the clones with higher proportions in the sample.

All clonal diversity plots and tables with diversity values can be found under `repertoire_analysis/Diversity`.
To correct for the different number of sequences in each of the samples, the Bootstrapping technique was employed,
in which `r nboot` random bootstrap samples were taken, with size the number of sequences in the sample with less sequences (N).
The solid line shows the mean Diversity of the bootstrap samples, whereas the transparent area shows the full Diversity
range of the bootstrap samples.


## Clonal diversity per subject
```{r clonal_diversity, echo = FALSE, include=FALSE}
sample_div <- alphaDiversity(abund, group="sample", min_q=0, max_q=4, step_q=0.05,
                            ci=0.95, nboot=nboot)
sample_main <- paste0("Sample diversity (N=", sample_div@n[1], ")")
sample_div@diversity$treatment <- sapply(sample_div@diversity$sample, function(x) unlist(strsplit(as.character(x), "_"))[1])
sample_div@diversity$time_point <- sapply(sample_div@diversity$sample, function(x) unlist(strsplit(as.character(x), "_"))[2])
sample_div@diversity$patient <- sapply(sample_div@diversity$sample, function(x) unlist(strsplit(as.character(x), "_"))[3])
```
```{r clonal_diversity2, fig.height=3, fig.width=12, echo = FALSE}
div_p <- ggplot(sample_div@diversity, aes(x = q, y = d,
                                        group = sample)) +
    geom_ribbon(aes(ymin = d_lower,
                    ymax = d_upper,
                    fill = time_point), alpha = 0.4) +
    geom_line(aes(color = time_point)) +
    xlab("q") + ylab(expression(""^q * D)) +
    ggtitle(sample_main) +
    facet_grid(cols=vars(patient))
div_p
```
```{r include = FALSE}
ggsave(plot=div_p, filename=paste0(diversity_dir,"/Diversity_patient_grid.png"), device="png", width = 25, height = 6, units="cm")
ggsave(plot=div_p, filename=paste0(diversity_dir,"/Diversity_patient_grid.pdf"), device="pdf", width = 25, height = 6, units="cm")
```

### Clonal diversity at specific q values {.tabset .tabset-fade .tabset-pills}

#### q=1 {-}

```{r clonal_diversity_q1, fig.height=4, fig.width=5, echo = FALSE}
sample_div@diversity$time_point <- factor(sample_div@diversity$time_point, levels=c("baseline","6months","12months"))
sample_div_q1 <- sample_div@diversity[which(sample_div@diversity$q == 1),]
sample_main <- paste0("Sample diversity at q=1 (N=", sample_div@n[1], ")")
dodge <- position_dodge(width = 0.9)
dq1 <- ggplot(sample_div_q1, aes(y=d, x=time_point)) +
    geom_point(aes(color=patient,group=time_point),position=dodge, stat="identity", size = 3) +
    #geom_errorbar(aes(ymin=d-d_sd, ymax=d+d_sd), width = .05, position=dodge) +
    geom_line(aes(group=patient,color=patient)) +
    xlab("") + ylab("Diversity (q=1)") +
    ggtitle(sample_main) +
    scale_colour_brewer(palette = "Accent")+
    #facet_grid(cols=vars(patient), drop=T, space="free", scales = "free") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
dq1
```

```{r, include = FALSE}
ggsave(plot = dq1, filename = paste0(diversity_dir,"/Diversity_q1_test_subject_paired.png"), device="png",
        width = 5, height = 4, units="cm")
ggsave(plot = dq1, filename = paste0(diversity_dir,"/Diversity_q1_test_subject_paired.pdf"), device="pdf",
        width = 5, height = 4, units="cm")
```

#### q=0 {-}

```{r clonal_diversity_q0, fig.height=4, fig.width=5, echo = FALSE}
sample_div_q0 <- sample_div@diversity[which(sample_div@diversity$q == 0),]
sample_main <- paste0("Sample diversity at q=0 (N=", sample_div@n[1], ")")
dodge <- position_dodge(width = 0.9)
dq0 <- ggplot(sample_div_q0, aes(y=d, x=time_point)) +
    geom_point(aes(color=patient,group=time_point),position=dodge, stat="identity", size = 3) +
    #geom_errorbar(aes(ymin=d-d_sd, ymax=d+d_sd), width = .05, position=dodge) +
    geom_line(aes(group=patient,color=patient)) +
    xlab("") + ylab("Diversity (q=0)") +
    ggtitle(sample_main) +
    scale_colour_brewer(palette = "Accent")+
    #facet_grid(cols=vars(patient), drop=T, space="free", scales = "free") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
dq0
```

```{r, include = FALSE}
ggsave(plot = dq0, filename = paste0(diversity_dir,"/Diversity_q0_test_subject_paired.png"), device="png",
        width = 25, height = 6, units="cm")
ggsave(plot = dq0, filename = paste0(diversity_dir,"/Diversity_q0_test_subject_paired.pdf"), device="pdf",
        width = 25, height = 6, units="cm")
```

## Clonal diversity per cell population

```{r clonal_diversity_pop, echo = FALSE, message=FALSE}
sample_div_pop <- alphaDiversity(abund_pop, group="sample_pop", min_q=0, max_q=4, step_q=0.05,
                                ci=0.95, nboot=nboot)
sample_main <- paste0("Sample diversity (N=", sample_div_pop@n[1], ")")
sample_div_pop@diversity$treatment <- sapply(sample_div_pop@diversity$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[1])
sample_div_pop@diversity$time_point <- sapply(sample_div_pop@diversity$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[2])
sample_div_pop@diversity$patient <- sapply(sample_div_pop@diversity$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[3])
sample_div_pop@diversity$population <- sapply(sample_div_pop@diversity$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[4])
```

```{r clonal_diversity_pop_plot, fig.width=12, echo=F, message=F}
sample_div_pop@diversity$time_point <- factor(sample_div_pop@diversity$time_point, levels = c("baseline","6months","12months"))
sample_div_pop@diversity$population <- factor(sample_div_pop@diversity$population, levels = c("N", "M", "DN", "P"))
sample_div_pop_df <-  sample_div_pop@diversity
div_pop <- ggplot(sample_div_pop@diversity, aes(x = q, y = d, group = sample_pop)) +
    geom_ribbon(aes(ymin = d_lower,
                    ymax = d_upper, fill = time_point), alpha = 0.4) +
    geom_line(aes(color = time_point)) +
    labs(color="Time Point", fill="Time Point") +
    xlab("q") + ylab(expression(""^q * D)) +
    ggtitle(sample_main) +
    facet_grid(cols=vars(patient), rows=vars(population), scales="free", drop = T)
div_pop
```

```{r, include=FALSE}
ggsave(plot = div_pop, filename = paste0(diversity_dir,"/Diversity_subject_population.png"), device="png",
        width = 27, height = 20, units="cm")
ggsave(plot = div_pop, filename = paste0(diversity_dir,"/Diversity_subject_population.pdf"), device="pdf",
        width = 27, height = 20, units="cm")
ggsave(plot = div_pop, filename = paste0(diversity_dir,"/Diversity_subject_population.png"), device="png",
        width = 27, height = 20, units="cm")
```

### Clonal diversity per population at specific q values {.tabset .tabset-fade .tabset-pills}

#### q=1 {-}

```{r clonal_diversity_q1_pop, echo = FALSE}
sample_div_q1 <- sample_div_pop@diversity[which(sample_div_pop@diversity$q == 1),]
sample_main <- paste0("Sample diversity at q=1 (N=", sample_div_pop@n[1], ")")
dodge <- position_dodge(width = 0.9)
dq1_pop <- ggplot(sample_div_q1, aes(y=d, x=time_point)) +
    geom_point(position=dodge, stat="identity", size = 1) +
    geom_errorbar(aes(ymin=d-d_sd, ymax=d+d_sd), width = .2, position=dodge) +
    xlab("") + ylab("Diversity (q=1)") +
    ggtitle(sample_main) +
    facet_grid(cols=vars(patient), rows=vars(population), drop=T, scales = "free") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
dq1_pop
```


```{r, include = FALSE}
ggsave(plot = dq1_pop, filename = paste0(diversity_dir,"/Diversity_q1_test_population.png"), device="png",
        width = 20, height = 15, units="cm")
ggsave(plot = dq1_pop, filename = paste0(diversity_dir,"/Diversity_q1_test_population.pdf"), device="pdf",
        width = 20, height = 15, units="cm")
```

```{r, fig.width=10, fig.asp=0.3,echo = FALSE}
sample_div_q1$subj_pop <- paste(sample_div_q1$patient, sample_div_q1$population, sep = "_")

sample_div_q1 <- sample_div_q1[!(sample_div_q1$sample_pop %in% c("Cladribin_12months_CLAD4_N","Cladribin_6months_CLAD4_N","Cladribin_baseline_CLAD4_N")),]

#
# boxplot_pop <- ggplot(sample_div_q1, aes(y=d, x=time_point)) +
#     geom_boxplot() +
#     geom_point(position=dodge, stat="identity", size = 1) +
#     #geom_errorbar(aes(ymin=d-d_sd, ymax=d+d_sd), width = .2, position=dodge) +
#     xlab("") + ylab("Diversity (q=1)") +
#     ggtitle(sample_main) +
#     facet_grid(cols=vars(population), drop=T, scales = "free") +
#     theme(axis.text.x = element_text(angle = 45, hjust = 1))
# boxplot_pop
#
# ggsave(plot = boxplot_pop, filename = paste0(diversity_dir,"/Diversity_q1_test_boxplot_population.png"), device="png",
#         width = 15, height = 6, units="cm")
# ggsave(plot = boxplot_pop, filename = paste0(diversity_dir,"/Diversity_q1_test_boxplot_population.pdf"), device="pdf",
#         width = 15, height = 6, units="cm")

# Boxplot with Wilcoxon paired tests (ggpubr)
my_comparisons <- list( c("baseline", "6months"), c("6months", "12months"), c("baseline", "12months") )

boxplot_divq1 <- ggboxplot(sample_div_q1, x = "time_point", y = "d",
        add = "jitter", legend = "none",
         color = "time_point", line.color = "gray", line.size = 0.4,
         facet.by = "population", ncol = 4, short.panel.labs = T) +
        scale_x_discrete(breaks=c("baseline","6months","12months"),
                      labels=c("B", "6M", "12M")) +
        #theme(axis.text.x = element_text(angle = 45, hjust=1), legend.position = "none") +
        xlab("") + ylab("Diversity (q=1)") +
        stat_compare_means(comparisons = my_comparisons, paired =T, method = "wilcox", label="p.signif", hide.ns = T) +
        stat_compare_means(label="p.format", label.y = 1300)

boxplot_divq1

boxplot_divq1_nokruskal <- ggboxplot(sample_div_q1, x = "time_point", y = "d",
        add = "jitter", legend = "none",
         color = "time_point", line.color = "gray", line.size = 0.4,
         facet.by = "population", ncol = 4, short.panel.labs = T) +
        scale_x_discrete(breaks=c("baseline","6months","12months"),
                      labels=c("B", "6M", "12M")) +
        #theme(axis.text.x = element_text(angle = 45, hjust=1), legend.position = "none") +
        xlab("") + ylab("Diversity (q=1)") +
        stat_compare_means(comparisons = my_comparisons, paired =T, method = "wilcox", label="p.signif", hide.ns = T)
boxplot_divq1_nokruskal
```

```{r, echo=F}
ggsave(plot=boxplot_divq1, filename = paste0(diversity_dir,"/Diversity_q1_subj_pop_boxplots.png"), device="png",
       width = 20, height = 7, units="cm")
ggsave(plot=boxplot_divq1, filename = paste0(diversity_dir,"/Diversity_q1_subj_pop_boxplots.pdf"), device="pdf",
       width=20, height=7, units="cm")
```

#### q=0 {-}

```{r clonal_diversity_q0_pop, fig.width=10, fig.asp = 0.6, echo = FALSE}
sample_div_q0 <- sample_div_pop@diversity[which(sample_div_pop@diversity$q == 0),]


sample_main <- paste0("Sample diversity at q=0 (N=", sample_div_pop@n[1], ")")
dodge <- position_dodge(width = 0.9)
dq0_pop <- ggplot(sample_div_q0, aes(y=d, x=time_point)) +
    geom_point(position=dodge, stat="identity", size = 1) +
    geom_errorbar(aes(ymin=d-d_sd, ymax=d+d_sd), width = .2, position=dodge) +
    xlab("") + ylab("Diversity (q=0)") +
    ggtitle(sample_main) +
    facet_grid(cols=vars(patient), rows=vars(population), drop=T, scales = "free") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
dq0_pop
```

```{r, include = FALSE}
ggsave(plot = dq0_pop, filename = paste0(diversity_dir,"/Diversity_q0_test_population.png"), device="png",
        width = 20, height = 15, units="cm")
ggsave(plot = dq0_pop, filename = paste0(diversity_dir,"/Diversity_q0_test_population.pdf"), device="pdf",
        width = 20, height = 15, units="cm")
```

```{r, fig.width=10, fig.asp=0.3,echo = FALSE}
# Boxplot with Wilcoxon paired tests (ggpubr)
sample_div_q0 <- sample_div_q0[!(sample_div_q0$sample_pop %in% c("Cladribin_12months_CLAD4_N","Cladribin_6months_CLAD4_N","Cladribin_baseline_CLAD4_N")),]

my_comparisons <- list( c("baseline", "6months"), c("6months", "12months"), c("baseline", "12months") )

boxplot_divq0 <- ggboxplot(sample_div_q0, x = "time_point", y = "d",
        add = "jitter", legend = "none",
         color = "time_point", line.color = "gray", line.size = 0.4,
         facet.by = "population", ncol = 4, short.panel.labs = T) +
        scale_x_discrete(breaks=c("baseline","6months","12months"),
                      labels=c("B", "6M", "12M")) +
        #theme(axis.text.x = element_text(angle = 45, hjust=1), legend.position = "none") +
        xlab("") + ylab("Diversity (q=0)") +
        stat_compare_means(comparisons = my_comparisons, paired =T, method = "wilcox", label="p.signif", hide.ns = T) +
        stat_compare_means(label="p.format", label.y=1300)
boxplot_divq0

boxplot_divq0_nokruskal <- ggboxplot(sample_div_q0, x = "time_point", y = "d",
        add = "jitter", legend = "none",
         color = "time_point", line.color = "gray", line.size = 0.4,
         facet.by = "population", ncol = 4, short.panel.labs = T) +
        scale_x_discrete(breaks=c("baseline","6months","12months"),
                      labels=c("B", "6M", "12M")) +
        #theme(axis.text.x = element_text(angle = 45, hjust=1), legend.position = "none") +
        xlab("") + ylab("Diversity (q=0)") +
        stat_compare_means(comparisons = my_comparisons, paired =T, method = "wilcox", label="p.signif", hide.ns = T)
boxplot_divq0_nokruskal
```

```{r, include = FALSE}
ggsave(plot = boxplot_divq0, filename = paste0(diversity_dir,"/Diversity_q0_test_population_boxplot.png"), device="png",
        width = 20, height = 7, units="cm")
ggsave(plot = boxplot_divq0, filename = paste0(diversity_dir,"/Diversity_q0_test_population_boxplot.pdf"), device="pdf",
        width = 20, height = 7, units="cm")
```

## Figure diversity and abundance metrics

```{r, fig.width=6, fig.asp=1,echo = FALSE}
all_boxplots <- cowplot::plot_grid(boxplot_divq0,
                          boxplot_divq1,
                          boxplot_clone_size,
          labels=c("A","B","C"),
          ncol=1,
          align = T
          )
all_boxplots
```

```{r, fig.width=6, echo=F, include=F}
ggsave(plot = all_boxplots, filename = paste0(diversity_dir,"/Figure_allboxplots.png"), device="png",
        width = 20, height = 16, units="cm")
ggsave(plot = all_boxplots, filename = paste0(diversity_dir,"/Figure_allboxplots.pdf"), device="pdf",
        width = 20, height = 16, units="cm")
```

```{r, fig.width=6, fig.asp=0.4,echo = FALSE}
all_boxplots_suppl <- plot_grid(boxplot_gini,
          labels=c("A"),
          ncol=1,
          align = T
          )
all_boxplots_suppl
```

```{r, fig.width=6, echo=F, include=F}
ggsave(plot = all_boxplots_suppl, filename = paste0(diversity_dir,"/Figure_allboxplots_suppl.png"), device="png",
        width = 20, height =7, units="cm")
ggsave(plot = all_boxplots_suppl, filename = paste0(diversity_dir,"/Figure_allboxplots_suppl.pdf"), device="pdf",
        width = 20, height = 7, units="cm")
```

```{r, fig.width=6, fig.asp=1,echo = FALSE}
all_boxplots_nokruskal <- plot_grid(
                          boxplot_divq0_nokruskal,
                          boxplot_divq1_nokruskal,
                          boxplot_clone_size_nokruskal,
          labels=c("A","B","C"),
          ncol=1,
          align = T
          )
all_boxplots_nokruskal
```

```{r, fig.width=6, echo=F, include=F}
ggsave(plot = all_boxplots_nokruskal, filename = paste0(diversity_dir,"/Figure_allboxplots_nokruskal.png"), device="png",
        width = 20, height = 16, units="cm")
ggsave(plot = all_boxplots_nokruskal, filename = paste0(diversity_dir,"/Figure_allboxplots_nokruskal.pdf"), device="pdf",
        width = 20, height = 16, units="cm")
```

```{r, fig.width=6, fig.asp=0.4,echo = FALSE}
all_boxplots_nokruskal_suppl <- plot_grid(boxplot_gini_nokruskal,
                          labels=c("A"),
          ncol=1,
          align = T
          )
all_boxplots_nokruskal_suppl
```

```{r, fig.width=6, echo=F, include=F}
ggsave(plot = all_boxplots_nokruskal_suppl, filename = paste0(diversity_dir,"/Figure_allboxplots_nokruskal_suppl.png"), device="png",
        width = 20, height = 7, units="cm")
ggsave(plot = all_boxplots_nokruskal, filename = paste0(diversity_dir,"/Figure_allboxplots_nokruskal_suppl.pdf"), device="pdf",
        width = 20, height = 7, units="cm")
```

# V gene usage

## V gene family usage {.tabset .tabset-fade .tabset-pills}

The V gene usage (in percentage) in each of the samples is represented below. All plots and tables can be found [here](./repertoire_comparison/V_family/).

Gene family usage is normalized by the number of clones.

### By patient {-}

```{r, echo=F,fig.width=10, fig.asp=0.3, fig.align='center'}
family <- countGenes(df_all, gene="v_call", groups="sample", mode="family", clone="clone_id")
family$treatment <- sapply(family$sample, function(x) unlist(strsplit(as.character(x), "_"))[1])
family$time_point <- sapply(family$sample, function(x) unlist(strsplit(as.character(x), "_"))[2])
family$time_point <- factor(family$time_point, levels=c("baseline","6months","12months"))
family$patient <- sapply(family$sample, function(x) unlist(strsplit(as.character(x), "_"))[3])


gg1 <- ggplot(family, aes(fill=gene, y=clone_freq, x=time_point)) +
        geom_bar(position="fill", stat="identity") +
        facet_grid(cols = vars(patient), drop=T, scales = "free" ) +
            theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1))


# g2 <- ggplot(family, aes(x=time_point, y=clone_freq, fill=time_point)) +
#     geom_bar(position = "dodge", stat="identity") +
#     ggtitle("V Gene Family Usage") +
#     theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1)) +
#     ylab("Frequency") +
#     xlab("") +
#     facet_grid(cols = vars(patient), rows=vars(gene),drop=T, scales = "free" ) +
#     theme(legend.position = "none") +
#     scale_fill_brewer(palette="Blues")
gg1
#g2
```

```{r, echo=F}
ggsave(filename = paste0(vfamily_dir, "/V_Family_distribution_patient.svg"), plot = gg1, width = 20, height = 6, units = "cm")
ggsave(filename = paste0(vfamily_dir, "/V_Family_distribution_patient.png"), plot = gg1, width = 20, height = 6, units = "cm")
write.table(family, file = paste0(vfamily_dir, "/V_family_distribution_data.tsv"), sep = "\t", quote = F, row.names = F)
```

```{r, echo=F,fig.width=10, fig.asp=0.3, fig.align='center'}
# dodge <- position_dodge(width = 0.9)
# g3 <- ggplot(family, aes(x=time_point, y=clone_freq)) +
#     geom_point(position =dodge, stat="identity", size=1) +
#     geom_boxplot() +
#     #geom_line(aes(group=patient,color=patient)) +
#     ggtitle("V Gene Family Usage") +
#     theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1)) +
#     ylab("Frequency") +
#     xlab("") +
#     facet_grid(cols=vars(gene),drop=T, scales = "free" )
#     #theme(legend.position = "none") +
#     #scale_colour_brewer(palette="Accent")
# g3


vgene_boxplot <- ggpubr::ggboxplot(family, x="time_point", y="clone_freq",
                        add = "jitter", legend="none", color = "time_point", line.color = "gray", line.size = 0.4,
                        facet.by = "gene", nrow = 1, short.panel.labs = T) +
                        scale_x_discrete(breaks=c("baseline","6months","12months"),
                      labels=c("B", "6M", "12M")) +
                      xlab("") + ylab("Frequency") +
                      stat_compare_means(comparisons = my_comparisons, paired=T, method = "wilcox", label="p.signif") +
                      stat_compare_means(paired=T, label="p.format", label.y = 0.9)
vgene_boxplot

```

```{r, echo=F}
ggsave(plot = vgene_boxplot, filename = paste0(vfamily_dir,"/V_Family_distribution_patient_boxplot.png"), device="png",
        width = 20, height = 7, units="cm")
ggsave(plot = vgene_boxplot, filename = paste0(vfamily_dir,"/V_Family_distribution_patient_boxplot.pdf"), device="pdf",
        width = 20, height = 7, units="cm")
```

### By Population {-}

```{r, echo=F, fig.width=10, fig.asp=0.8, fig.align='center'}
family_pop <- countGenes(df_all, gene="v_call", groups="sample_pop",
                    mode="family", clone = "clone_id")
family_pop$treatment <- sapply(family_pop$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[1])
family_pop$time_point <- sapply(family_pop$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[2])
family_pop$patient <- sapply(family_pop$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[3])
family_pop$population <- sapply(family_pop$sample_pop, function(x) unlist(strsplit(as.character(x), "_"))[4])
family_pop$time_point <- factor(family_pop$time_point, levels=c("baseline","6months","12months"))
family_pop$population <- factor(family_pop$population, levels=c("N","M","DN","P"))


gg1 <- ggplot(family_pop, aes(fill=gene, y=clone_freq, x=time_point)) +
        geom_bar(position="fill", stat="identity") +
        facet_grid(cols = vars(patient), rows = vars(population), drop=T, scales = "free" ) +
        scale_x_discrete(breaks=c("baseline","6months","12months"),
                      labels=c("B", "6M", "12M")) +
        xlab("") + ylab("Frequency")

gg1

```

```{r, echo = F}
ggsave(filename = paste0(vfamily_dir,"/V_Family_distribution_patient_population.svg"), plot = gg1,
    width = 20, height = 16, units = "cm")
ggsave(filename = paste0(vfamily_dir,"/V_Family_distribution_patient_population.png"), plot = gg1,
    width = 20, height = 16, units = "cm")
ggsave(filename = paste0(vfamily_dir,"/V_Family_distribution_patient_population.pdf"), plot = gg1,
    width = 20, height = 16, units = "cm")
write.table(family, file = paste0(vfamily_dir, "/V_family_distribution_data_population.tsv"), sep = "\t",
    quote = F, row.names = F)
```

```{r, echo=F,fig.width=10, fig.asp=1, fig.align='center'}
#family_pop_m <- family_pop[(family_pop$population=="M"),]

family_pop <- family_pop[!(family_pop$sample_pop %in% c("Cladribin_12months_CLAD4_N","Cladribin_6months_CLAD4_N","Cladribin_baseline_CLAD4_N")),]

# dodge <- position_dodge(width = 0.9)
# g3 <- ggplot(family_pop, aes(x=time_point, y=clone_freq)) +
#       geom_boxplot() +
#     geom_point(position=dodge, stat="identity", size=1) +
#
#     #geom_line(aes(group=patient,color=patient)) +
#     ggtitle("V Gene Family Usage") +
#     theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1)) +
#     ylab("Frequency") +
#     xlab("") +
#     facet_grid(cols=vars(gene), rows=vars(population), drop=T, scales = "free" )
#     #theme(legend.position = "none") +
#     #scale_colour_brewer(palette="Accent")
# g3

my_comparisons <- list( c("baseline", "6months"), c("6months", "12months"), c("baseline", "12months") )

# Remove IGHV7 as no primers were included for this gene:
family_pop <- family_pop %>% filter(gene != "IGHV7")

vgene_boxplot_pop <- ggpubr::ggboxplot(family_pop, x="time_point", y="clone_freq",
                        add = "jitter", legend="none", color = "time_point", line.color = "gray", line.size = 0.4,
                        facet.by = c("population","gene"), short.panel.labs = T) +
                        scale_x_discrete(breaks=c("baseline","6months","12months"),
                      labels=c("B", "6M", "12M")) +
                      xlab("") + ylab("Frequency") +
                      stat_compare_means(comparisons = my_comparisons, paired=T, method = "wilcox", label="p.signif", hide.ns = T) +
                      stat_compare_means(paired=T, label="p.format", label.y = 1.1, hide.ns = T)
vgene_boxplot_pop

```

```{r, echo=F}
ggsave(plot = vgene_boxplot_pop, filename = paste0(vfamily_dir,"/V_gene_family_usage_boxplot_population.png"), device="png",
        width = 20, height =20, units="cm")
ggsave(plot = vgene_boxplot_pop, filename = paste0(vfamily_dir,"/V_gene_family_usage_boxplot_population.pdf"), device="pdf",
        width = 20, height =20, units="cm")

```

## V gene usage {.tabset .tabset-fade .tabset-pills}

The V gene usage (in percentage) in each of the samples is represented below.
All plots and tables can be found [here](./repertoire_comparison/V_family/).

### By clones {-}

```{r, echo=F, fig.width=10, fig.asp=1.8, fig.align='center'}
family <- countGenes(df_all, gene="v_call", groups="sample", mode="gene", clone = "clone_id")
family$treatment <- sapply(family$sample, function(x) unlist(strsplit(as.character(x), "_"))[1])
family$time_point <- sapply(family$sample, function(x) unlist(strsplit(as.character(x), "_"))[2])
family$patient <- sapply(family$sample, function(x) unlist(strsplit(as.character(x), "_"))[3])
g2 <- ggplot(family, aes(x=gene, y=clone_freq, fill=gene)) +
    scale_color_brewer(palette="Set1") +
    geom_bar(position = "dodge", stat="identity") +
    ggtitle("V Gene Family Usage") +
    theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1)) +
    ylab("Frequency") +
    xlab("") +
    facet_wrap(vars(patient), scales="free_x", ncol = 1) +
    theme(legend.position = "none")
g2
ggsave(filename = paste0(vfamily_dir, "/V_gene_distribution_by_clone_patient.svg"), plot = g2, width = 20, height = 40, units = "cm")
ggsave(filename = paste0(vfamily_dir, "/V_gene_distribution_by_clone_patient.png"), plot = g2, width = 20, height = 40, units = "cm")
write.table(family, file = paste0(vfamily_dir, "/V_gene_distribution_by_clone_data.tsv"), sep = "\t", quote = F, row.names = F)
```

### By sequences {-}

```{r, echo=F, fig.width=10, fig.asp=1.8, fig.align='center'}
family <- countGenes(df_all, gene="v_call", groups="sample", mode="gene")
family$treatment <- sapply(family$sample, function(x) unlist(strsplit(as.character(x), "_"))[1])
family$time_point <- sapply(family$sample, function(x) unlist(strsplit(as.character(x), "_"))[2])
family$patient <- sapply(family$sample, function(x) unlist(strsplit(as.character(x), "_"))[3])
g2 <- ggplot(family, aes(x=gene, y=seq_freq, fill=gene)) +
    scale_color_brewer(palette="Set1") +
    geom_bar(position = "dodge", stat="identity") +
    ggtitle("V Gene Usage") +
    theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1)) +
    ylab("Frequency") +
    xlab("") +
    facet_wrap(vars(patient), scales="free_x", ncol = 1) +
    theme(legend.position = "none")
g2
ggsave(filename = paste0(vfamily_dir, "/V_gene_distribution_by_sequence_patient.svg"), plot = g2, width = 20, height = 40, units = "cm")
ggsave(filename = paste0(vfamily_dir, "/V_gene_distribution_by_sequence_patient.png"), plot = g2, width = 20, height = 40, units = "cm")
write.table(family, file = paste0(vfamily_dir, "/V_gene_distribution_by_sequence_data.tsv"), sep = "\t", quote = F, row.names = F)
```

# Isotype usage

## Isotype usage per subject

```{r isotype_usage, echo=F, fig.width=10, fig.asp=0.3, fig.align='center'}
# Plotting Isotype percentages per patient
df_all$isotype <- df_all$c_primer

res <- df_all %>% group_by(isotype,sample,source,treatment,extract_time) %>% dplyr::summarise(Seqs_isotype=n())
res <- with(res, res[order(source),])
res_sample <- df_all %>% group_by(sample) %>% dplyr::summarise(Seqs_total=n())

freqs <- merge(x=res, y=res_sample, all.x = T, by.x = "sample", by.y = "sample")
freqs$Freq <- (freqs$Seqs_isotype/freqs$Seqs_total)
freqs$extract_time <- factor(freqs$extract_time, levels=c("baseline","6months","12months"))

# g4 <- ggplot(freqs, aes(fill=extract_time, y=Freq, x=isotype)) +
#   geom_bar(position = "dodge", stat="identity") +
#   xlab("") +
#   ylab("Frequency") +
#   ggtitle("Isotype frequency") +
#   facet_grid(cols=vars(source), scales = "free", drop = T) +
#   theme(axis.text.x = element_text(angle=45, hjust = 1, vjust = 1))
# g4

g4 <- ggplot(freqs, aes(fill=isotype, y=Freq, x=extract_time)) +
        geom_bar(position="fill", stat="identity") +
        facet_grid(cols = vars(source), drop=T, scales = "free" ) +
            theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1))
g4
```

```{r, echo=F}
ggsave(plot=g4, filename = paste0(isotype_dir,"/Isotype_frequencies_patient.svg"), device = "svg",
  width = 20, height = 6, units = "cm")
ggsave(plot=g4, filename = paste0(isotype_dir,"/Isotype_frequencies_patient.pdf"), device = "pdf",
  width = 20, height = 6, units = "cm")
write.table(freqs, file = paste0(isotype_dir,"/Isotype_frequencies_data.tsv"), sep="\t", quote=F, row.names = F)

```

## Isotype usage per cell population

```{r isotype_pop, echo=F, fig.width=15, fig.asp=0.6, fig.align='center'}
df_all$isotype <- df_all$c_primer
res_pop <- df_all %>% group_by(isotype, sample_pop, source, treatment, extract_time, population) %>% dplyr::summarise(Seqs_isotype=n())
res_pop <- with(res_pop, res_pop[order(source),])
res_pop_sample <- df_all %>% group_by(sample_pop) %>% dplyr::summarise(Seqs_total=n())

freqs <- merge(x=res_pop, y=res_pop_sample, all.x = T, by.x = "sample_pop", by.y = "sample_pop")
freqs$Freq <- (freqs$Seqs_isotype/freqs$Seqs_total)
freqs$extract_time <- factor(freqs$extract_time, levels=c("baseline","6months","12months"))
freqs$population <- factor(freqs$population, levels=c("N", "M", "DN", "P"))
# g4 <- ggplot(freqs, aes(fill=extract_time, y=Freq, x=isotype)) +
#  geom_bar(position = "dodge", stat="identity") +
#  xlab("") +
#  ylab("Frequency") +
#  ggtitle("Isotype frequency") +
#  facet_grid(cols=vars(source), rows=vars(population)) +
#  theme(axis.text.x = element_text(angle=45, hjust = 1, vjust = 1))
# g4

g4 <- ggplot(freqs, aes(fill=isotype, y=Freq, x=extract_time)) +
        geom_bar(position="fill", stat="identity") +
        facet_grid(cols = vars(source), rows=vars(population),drop=T, scales = "free" ) +
          scale_x_discrete(breaks=c("baseline","6months","12months"),
                      labels=c("B", "6M", "12M"))
g4
```

```{r, echo=F}
ggsave(g4, filename = paste0(isotype_dir,"/Isotype_percentages_cell_population.svg"), device = "svg",
  width = 20, height = 14, units = "cm")
ggsave(g4, filename = paste0(isotype_dir,"/Isotype_percentages_cell_population.pdf"), device = "pdf",
  width = 20, height = 14, units = "cm")
ggsave(g4, filename = paste0(isotype_dir,"/Isotype_percentages_cell_population.png"), device = "png",
  width = 20, height = 14, units = "cm")

write.table(freqs, file = paste0(isotype_dir, "/Isotype_frequencies_cell_population_data.tsv"), sep="\t", quote = F, row.names = F)
```

```{r, echo=F, fig.width=10, fig.asp=1}
my_comparisons <- list( c("baseline", "6months"), c("6months", "12months"), c("baseline", "12months") )

freqs <- freqs[!(freqs$sample_pop %in% c("Cladribin_12months_CLAD4_N","Cladribin_6months_CLAD4_N","Cladribin_baseline_CLAD4_N")),]

isotype_boxplot_pop <- ggpubr::ggboxplot(freqs, x="extract_time", y="Freq",
                        add = "jitter", legend="none", color = "extract_time", line.color = "gray", line.size = 0.4,
                        facet.by = c("population","isotype"), short.panel.labs = T) +
                        scale_x_discrete(breaks=c("baseline","6months","12months"),
                      labels=c("B", "6M", "12M")) +
                      xlab("") + ylab("Frequency") +
                      stat_compare_means(comparisons = my_comparisons, paired=T, method = "wilcox", label="p.signif", hide.ns=T) +
                      stat_compare_means(paired=T, label="p.format", label.y = 1.4, hide.ns = T)
isotype_boxplot_pop
```

```{r, echo=F}
ggsave(plot = isotype_boxplot_pop, filename = paste0(isotype_dir,"/Isotype_usage_boxplot_population.png"), device="png",
        width = 20, height =20, units="cm")
ggsave(plot = isotype_boxplot_pop, filename = paste0(isotype_dir,"/Isotype_usage_boxplot_population.pdf"), device="pdf",
        width = 20, height =20, units="cm")

```

```{r, echo=F,fig.width=10, fig.asp=0.3, fig.align='center'}
isotype_pop_m <- freqs[(freqs$population=="M"),]

isotype_boxplot_pop_m <- ggpubr::ggboxplot(isotype_pop_m, x="extract_time", y="Freq",
                        add = "jitter", legend="none", color = "extract_time", line.color = "gray", line.size = 0.4,
                        facet.by = c("population","isotype"), short.panel.labs = T) +
                        scale_x_discrete(breaks=c("baseline","6months","12months"),
                      labels=c("B", "6M", "12M")) +
                      xlab("") + ylab("Frequency") +
                      stat_compare_means(comparisons = my_comparisons, paired=T, method = "wilcox", label="p.signif") +
                      stat_compare_means(paired=T, label="p.format", label.y = 1.4)
isotype_boxplot_pop_m
```

```{r, echo=F}
ggsave(plot = isotype_boxplot_pop_m, filename = paste0(isotype_dir,"/Isotype_usage_boxplot_M_population.png"), device="png",
        width = 15, height = 6, units="cm")
ggsave(plot = isotype_boxplot_pop_m, filename = paste0(isotype_dir,"/Isotype_usage_boxplot_M_population.pdf"), device="pdf",
        width = 15, height = 6, units="cm")
```




# Citations

If you use nf-core/bcellmagic for your analysis, please cite it using the following DOI: [10.5281/zenodo.3607408](https://doi.org/10.5281/zenodo.3607408)

Please also cite the `nf-core` publication [@ewels2020].

In addition, citations for the tools and data used in this pipeline are as follows:

* **pRESTO** [@VanderHeiden2014]
* **SHazaM, Change-O** [@Gupta2015]
* **Alakazam** [@Stern2014]
* **TIgGER** [@Gadala-Maria2015]
* **FastQC** [@andrews2010]
* **MultiQC** [@ewels2016]
